drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2163 @@
+	 * received
+	 */
+	LL_I3C_DisableIT_TXFNF(i3c);
+
+	/* Start DAA */
+	LL_I3C_ControllerHandleCCC(i3c, I3C_CCC_ENTDAA, 0, LL_I3C_GENERATE_STOP);
+
+	/* Wait for DAA to finish */
+	if (k_sem_take(&data->device_sync_sem, STM32_I3C_TRANSFER_TIMEOUT) != 0) {
+		return -ETIMEDOUT;
+	}
+
+	if (data->msg_state == STM32_I3C_MSG_ERR) {
+		i3c_stm32_clear_err(dev, false);
+		/* Enable TXFNF interrupt in case an error occurred before it was enabled by RXFNE
+		 */
+		LL_I3C_EnableIT_TXFNF(i3c);
+		return -EIO;
+	}
+
+	k_mutex_unlock(&data->bus_mutex);
+
+	return 0;
+}
+
+#ifdef CONFIG_I3C_STM32_DMA
+
+static int i3c_stm32_dma_msg_control_fifo_config(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	int ret;
+
+	data->dma_tc.blk_cfg.source_address = (uint32_t)data->control_fifo;
+	data->dma_tc.blk_cfg.block_size = data->fifo_len;
+
+	ret = dma_config(data->dma_tc.dma_dev, data->dma_tc.dma_channel, &data->dma_tc.dma_cfg);
+
+	if (ret != 0) {
+		LOG_ERR("Control DMA config error, err=%d", ret);
+		return -EINVAL;
+	}
+
+	if (dma_start(data->dma_tc.dma_dev, data->dma_tc.dma_channel)) {
+		LOG_ERR("Control DMA start failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int i3c_stm32_dma_msg_status_fifo_config(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	int ret;
+
+	data->dma_rs.blk_cfg.dest_address = (uint32_t)data->status_fifo;
+	data->dma_rs.blk_cfg.block_size = data->fifo_len;
+
+	ret = dma_config(data->dma_rs.dma_dev, data->dma_rs.dma_channel, &data->dma_rs.dma_cfg);
+
+	if (ret != 0) {
+		LOG_ERR("Status DMA config error, err=%d", ret);
+		return -EINVAL;
+	}
+
+	if (dma_start(data->dma_rs.dma_dev, data->dma_rs.dma_channel)) {
+		LOG_ERR("Status DMA start failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static int i3c_stm32_dma_msg_config(const struct device *dev, uint32_t buf_addr, size_t buf_len)
+{
+	struct i3c_stm32_dma_stream *dma_stream;
+	struct i3c_stm32_data *data = dev->data;
+	int ret;
+
+	if (i3c_stm32_curr_msg_xfer_is_read(dev)) {
+		dma_stream = &(data->dma_rx);
+		dma_stream->blk_cfg.dest_address = buf_addr;
+	} else {
+		dma_stream = &(data->dma_tx);
+		dma_stream->blk_cfg.source_address = buf_addr;
+	}
+
+	i3c_stm32_arbitration_header_config(dev);
+
+	dma_stream->blk_cfg.block_size = buf_len;
+	ret = dma_config(dma_stream->dma_dev, dma_stream->dma_channel, &dma_stream->dma_cfg);
+
+	if (ret != 0) {
+		LOG_ERR("TX/RX DMA config error, err=%d", ret);
+		return -EINVAL;
+	}
+
+	if (dma_start(dma_stream->dma_dev, dma_stream->dma_channel)) {
+		LOG_ERR("TX/RX DMA start failed");
+		return -EFAULT;
+	}
+	return 0;
+}
+#endif
+
+static int i3c_stm32_transfer_begin(const struct device *dev)
+{
+	struct i3c_stm32_data *data = dev->data;
+	const struct i3c_stm32_config *config = dev->config;
+	I3C_TypeDef *i3c = config->i3c;
+
+	data->msg_state = STM32_I3C_MSG;
+	data->sf_state = STM32_I3C_SF;
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
```

Comment:
nit: these #ifdef are not neccessary around `pm_device_runtime_*` as it will look at that kconfig within the function